azkaban-common/src/main/java/azkaban/storage/HdfsAuth.java

```diff
@@ -53,25 +58,47 @@ public HdfsAuth(Props props, Configuration conf) {
   }
 
   /**
+   * Wrapper for Hadoop doAs
+   *
+   * All Hadoop/HDFS access calls must be wrapped in a doAs call. This utility automatically
+   * logs in the user/refreshes the TGT and performs action.run()
+   *
+   * @param action executable block of code which is to be run in the doAs scope
+   * @param <T> return class type
+   * @return object returned by action.run()
+   */
+  public <T> T doAs(PrivilegedExceptionAction<T> action) {
+    authorize();
+    try {
+      return requireNonNull(loggedInUser).doAs(action);
+    } catch (Exception e) {
+      throw new StorageException(
+          String.format("Error while performing doAs user: %s", loggedInUser), e);
+    }
+  }
+
+  /**
    * API to authorize HDFS access.
    * This logins in the configured user via the keytab.
    * If the user is already logged in then it renews the TGT.
    */
-  public void authorize() {
+  private void authorize() {
     if (isSecurityEnabled) {
       try {
         login(keytabPrincipal, keytabPath);
       } catch (IOException e) {
         log.error(e);
         throw new AzkabanException(String.format(
-            "Error: Unable to authorize to Hadoop. Principal: %s Keytab: %s", keytabPrincipal, keytabPath));
+            "Error: Unable to authorize to Hadoop. Principal: %s Keytab: %s", keytabPrincipal,
+            keytabPath));
       }
     }
   }
 
   private void login(String keytabPrincipal, String keytabPath) throws IOException {
     if (loggedInUser == null) {
```

Comment:
The loggedInUser is not volatile and the method is not synchronized, the check is not guaranteed to work.